compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -270,45 +270,43 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
 
 pub(super) fn elf_e_flags(architecture: Architecture, sess: &Session) -> u32 {
     match architecture {
-        Architecture::Mips => {
-            let arch = match sess.target.options.cpu.as_ref() {
+        Architecture::Mips | Architecture::Mips64 | Architecture::Mips64_N32 => {
+            let mut e_flags = match sess.target.options.cpu.as_ref() {
                 "mips1" => elf::EF_MIPS_ARCH_1,
                 "mips2" => elf::EF_MIPS_ARCH_2,
                 "mips3" => elf::EF_MIPS_ARCH_3,
                 "mips4" => elf::EF_MIPS_ARCH_4,
                 "mips5" => elf::EF_MIPS_ARCH_5,
-                s if s.contains("r6") => elf::EF_MIPS_ARCH_32R6,
+                "mips32r2" => elf::EF_MIPS_ARCH_32R2,
+                "mips32r6" => elf::EF_MIPS_ARCH_32R6,
+                "mips64r2" => elf::EF_MIPS_ARCH_64R2,
+                "mips64r6" => elf::EF_MIPS_ARCH_64R6,
                 _ => elf::EF_MIPS_ARCH_32R2,
             };
 
-            let mut e_flags = elf::EF_MIPS_CPIC | arch;
-
-            // If the ABI is explicitly given, use it or default to O32.
+            // If the ABI is explicitly given, use it or default to O32 on 32-bit MIPS
+            // and N64 (which needs no extra flag) on 64-bit MIPS.
             match sess.target.options.llvm_abiname.to_lowercase().as_str() {
                 "n32" => e_flags |= elf::EF_MIPS_ABI2,
                 "o32" => e_flags |= elf::EF_MIPS_ABI_O32,
-                _ => e_flags |= elf::EF_MIPS_ABI_O32,
+                _ => {
+                    if architecture == Architecture::Mips {
+                        e_flags |= elf::EF_MIPS_ABI_O32
+                    }
+                }
```

Comment:
I do not like this continued guessing. Either
- do nothing on this branch
- explicitly recognize `"n64"` as a possible variant, add it to all the 64-bit MIPS targets, and panic if `llvm_abiname` is unset

Reply:
I am at about the point where I think we should require the field on all MIPSy targets, just like it is required on RISCV (last I checked) and a similar-but-different field is required on x86. Especially for ABI-affecting fields, we generally make fields required and then. It's basically the same thing as the lowercase issue: if there is a sensible default for or normalization of the data, then we should probably address that during deserialization and emit a warning if the field was not specified.

Or just panic on the field being unspecified, either one.

Reply:
I am okay with the 32-bit case, I think, being inferred, due to LLVM not supporting other alternative MIPS ABIs for 32-bit MIPS (though supposedly they exist)